proto/pulumi/provider.proto

```diff
@@ -363,81 +373,136 @@ message CheckFailure {
+//   field to further communicate which changes (if any) require a replacement as opposed to an update.
+//
+// * *Detailed diffs* are those with `hasDetailedDiff` set, and utilise the `detailedDiff` field to provide a more
+//   granular view of the changes that have occurred. Detailed diffs are designed to allow providers to control
+//   precisely which field names are displayed as responsible for a change, and to signal more accurately what kind of
+//   change occurred (e.g. a field was added, deleted or updated).
 message DiffResponse {
-    repeated string replaces = 1; // if this update requires a replacement, the set of properties triggering it.
-    repeated string stables = 2;  // an optional list of properties that will not ever change.
-    bool deleteBeforeReplace = 3; // if true, this resource must be deleted before replacing it.
-    DiffChanges changes = 4;      // if true, this diff represents an actual difference and thus requires an update.
+    // A set of properties which have changed and whose changes require the resource being diffed to be replaced. The
+    // caller should replace the resource if this set is non-empty, or if any of the properties specified in
+    // `detailedDiff` have a `*_REPLACE` kind.
+    repeated string replaces = 1;
+
+    // An optional list of properties that will not ever change (are stable).
+    repeated string stables = 2;
+
+    // If true, this resource must be deleted *before* its replacement is created.
+    bool deleteBeforeReplace = 3;
+
+    // The result of the diff. Indicates at a high level whether the resource has changed or not (or, in legacy cases,
+    // if the provider does not know).
+    DiffChanges changes = 4;
 
-    // a list of the properties that changed. This should only contain top level property names, it does not
-    // support nested properties. For that use detailedDiff.
+    // The set of properties which have changed. This field only supports top-level properties. It *does not* support
+    // full [property paths](property-paths); implementations should use `detailedDiff` when this is required.
     repeated string diffs = 5;
 
-    // detailedDiff is an optional field that contains map from each changed property to the type of the change.
-    //
-    // The keys of this map are property paths. These paths are essentially Javascript property access expressions
-    // in which all elements are literals, and obey the following EBNF-ish grammar:
-    //
-    //   propertyName := [a-zA-Z_$] { [a-zA-Z0-9_$] }
-    //   quotedPropertyName := '"' ( '\' '"' | [^"] ) { ( '\' '"' | [^"] ) } '"'
-    //   arrayIndex := { [0-9] }
-    //
-    //   propertyIndex := '[' ( quotedPropertyName | arrayIndex ) ']'
-    //   rootProperty := ( propertyName | propertyIndex )
-    //   propertyAccessor := ( ( '.' propertyName ) |  propertyIndex )
-    //   path := rootProperty { propertyAccessor }
-    //
-    // Examples of valid keys:
```

Comment:
Is this captured in the dev docs already? Here it's marked as todo: https://pulumi-developer-docs.readthedocs.io/latest/developer-docs/providers/implementers-guide.html#property-paths

Reply:
There's an updated section at https://pulumi-developer-docs.readthedocs.io/latest/docs/architecture/types/README.html#property-paths now, which I think works? But that needs fleshing out too, for sure.